modules/apigee_edge_apidocs/tests/src/Functional/ApiDocsAccessTest.php

```diff
@@ -0,0 +1,418 @@
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+namespace Drupal\Tests\apigee_edge_apidocs\Functional;
+
+use Drupal\apigee_edge_apidocs\Entity\ApiDoc;
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Tests the ApiDoc term access permissions.
+ *
+ * @group apigee_edge
+ * @group apigee_edge_apidocs
+ */
+class ApiDocsAccessTest extends BrowserTestBase {
+
+  /**
+   * Modules to enable.
+   *
+   * @var array
+   */
+  public static $modules = ['apigee_edge_apidocs', 'block', 'field_ui'];
+
+  /**
+   * A published API Doc.
+   *
+   * @var \Drupal\apigee_edge_apidocs\Entity\ApiDoc
+   */
+  protected $apidocPublished;
+
+  /**
+   * An unpublished API Doc.
+   *
+   * @var \Drupal\apigee_edge_apidocs\Entity\ApiDoc
+   */
+  protected $apidocUnpublished;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+
+    // Add the system menu blocks to appropriate regions.
+    $this->setupMenus();
+
+    // Create published apidoc.
+    $apidoc_published = ApiDoc::create([
+      'name' => 'API 1',
+      'description' => 'Test API 1',
+      'spec' => NULL,
+      'status' => 1,
+    ]);
+    $apidoc_published->save();
+    $this->apidocPublished = $apidoc_published;
+
+    // Create unpublished apidoc.
+    $apidoc_unpublished = ApiDoc::create([
+      'name' => 'API 2',
+      'description' => 'Test API 2',
+      'spec' => NULL,
+      'status' => 0,
+    ]);
+    $apidoc_unpublished->save();
+    $this->apidocUnpublished = $apidoc_unpublished;
+  }
+
+  /**
+   * Set up menus and tasks in their regions.
+   *
+   * Since menus and tasks are now blocks, we're required to explicitly set them
+   * to regions.
+   */
+  protected function setupMenus() {
+    $this->drupalPlaceBlock('system_menu_block:tools', ['region' => 'primary_menu']);
+    $this->drupalPlaceBlock('local_tasks_block', ['region' => 'secondary_menu']);
+    $this->drupalPlaceBlock('local_actions_block', ['region' => 'content']);
+    $this->drupalPlaceBlock('page_title_block', ['region' => 'content']);
+  }
+
+  /**
+   * Test admin access control functionality for apidocs.
+   */
+  public function testApiDocAccessAdmin() {
+    $assert_session = $this->assertSession();
+
+    // Test the 'administer apidoc entities' permission.
+    $this->drupalLogin($this->drupalCreateUser([
+      'administer apidoc entities',
+      'administer apidoc display',
+      'administer apidoc fields',
+      'administer apidoc form display',
+    ]));
+
+    $this->drupalGet('apidoc/' . $this->apidocPublished->id());
```

Comment:
Always generate URLs from routes.

Reply:
Updated by dawehner to use the url generated by the entities instead: `$this->drupalGet($this->apidocPublished->toUrl());`